### ISSUE TITLE
Burning backups to DVD

### ISSUE THREAD

**Issue by shasheene** [rescuezilla/rescuezilla#9]
There have been a [couple](https://sourceforge.net/p/redobackup/discussion/1169664/thread/bf84a441/) [of](https://sourceforge.net/p/redobackup/discussion/1169663/thread/fcd14731/) [threads](https://sourceforge.net/p/redobackup/discussion/1169664/thread/b84fcbe1/) from 2010/2011 calling for the ability to backup a drive onto a DVD.

This may be achieved by placing optical media burning software in Rescuezilla.

There are two approaches:
1. Relying on users to first backup to an existing hard drive before using burning software
1. Burning directly to disk without first backing up to an existing hard drive

Both approaches would require restore logic to handle a single backup separated over multiple disks.

Either approach is likely to be a lot of work to implement. Few people burn optical media in 2019, even with 50GB and 100GB Blu-Ray recordable disks.

It would be good to support this, but it's low priority and given the small target market it is infeasible without sufficent [Patreon](https://www.patreon.com/rescuezilla) funding. Please comment/react if you would like this feature.

**Comment by AE720** [comment]
I understand the issues arising from something like this. IMO Disk 1 should have the real brains and some of a part while following disks are raw zips. The underlying concern is caching and where to do it.

On the top of my head, the easier and sloppier way is to find a way to cache it in RAM however that would only be even semi-feasible on single discs as expecting computers to have more than 4gb of ram (Would need more for Rescuezilla itself) would isolate the issue. A failcheck could be a message saying the system does not meet the required resources to perform this operation. Anything above a dual layer DVD, I'd imagine would just be able to treat the disk just like a shared drive however I'm also sure that it's more complicated to perform operations to a disc besides a straight burn. This wouldn't be a problem if we could use Window's Live FIle System. 

The "proper" approach would be to be able to, on the fly, know everything that's getting burned to the disk, have a middleman handle the task of making it look like it's a simple burn to the disc burning software and using the RAM for caching as it's getting processed, zipped and written to the disk in sequence.

To try and clarify, to the software burning to the disc, it's getting the whole part, but that information is spoofed as behind it, the archiving is done, bit by bit it moves from disk you're backing up, to the archive as archive is getting that content added to it on the disk. Obviously I'm not a programmer is this probably doesn't make a lot of sense and I'm making assumptions about what disc burning AND file archiving can and can't do. 

I don't even see any software that does anything similar like InfraRecorder. Those are my thoughts on it. It sounds like it wouldn't be worth the effort to do anything but throw in a program that burns to a disc from an already existing backup. I do know my home Debian server with all of the resources I'm running can fit on 3.7gb so I know I'd utilize it a bit. Optical storage seems to be the best thing we have at long term, cold, storage.

**Comment by pfrouleau** [comment]
My 2 cents:
- In the past, we had to feed the writer as fast as it was writing, otherwise, it was corrupting the disk. Because of this, I was always preparing my disks before burning them. I didn't buy or write any DVD for a long time. Hard drives are affordable now, we can reuse them when the data is not needed anymore, and USB3 is fast.
- In an 1st phase, Rescuezilla could support restoring for an optical disk. The user would be responsible to take the data from an HDD and burn the disks. To help the user, Rescuezilla could organize the files in subdirectories for each, but that would also increase the complexity of restoration because it would have to read from the subdirectories too. If the data is really on an optical disk, Rescuezilla should ask the user to insert the disk with the file XYZ.ac...
- In an 2nd phase, Rescuezilla could create the ISO themself. I do not remember if we can write directly to an ISO in Linux (as I said, did not do that for a long time). Maybe the split size could be set to the ISO size to create (less a few MB) and maybe the metadata files should be copied in each ISO.
- A feature to protect against bit rotting would be to add Parchive support. This can even help to protect data on optical disk against scratches. See https://en.wikipedia.org/wiki/Parchive. _I even made DVDs full of par2 files generated to rebuild DVDs in case they got scratched, but I never had to use them so far._

**Comment by shasheene** [comment]
Yes, almost all metadata files are no problem, because they are all generated right at the beginning of the backup operation, and the files tend to be very small (a few kilobytes):

```
Info-lshw.txt 17K
Info-dmi.txt 1.9K
lvm_cl.conf 1.7K
Info-lspci.txt 1.4K
Info-smart.txt 888
blkid.list 698
sdi-mbr 512
sdi-pt.parted 339
blkdev.list 336
sdi-pt.parted.compact 311
lvm_logv.list 212
dev-fs.list 203
sdi-pt.sf 192
Info-img-id.txt 186
Info-packages.txt 167
swappt-cl-swap.info 53
clonezilla-img 44K
sdi-hidden-data-after-mbr 1.0M
cl-root.xfs-ptcl-img.gz.aa 1.5G
sdi1.ext4-ptcl-img.gz.aa 183M
```

It's no problem leaving them in RAM until they're ready to burn. The only gotcha is the `clonezilla-img` log file is not complete until after the entire backup operation has completed. Side note: The "hidden-data-after-mbr" file up to 1 gigabyte in size if there's a big gap before the first partition, and 1GB of RAM usage is arguably large enough to warrant more advanced handling see below.

As you've said, many users may have eg, 8GB of RAM but the partitions they backup could be multiple terabytes. The Linux pipes used internally neatly handle this situation. Even holding the full 4.7GB data temporarily RAM could work, but I believe it's possible to do with only a few hundred megabyte buffer. Buffer underrun protection is quite mature, so making coasters out of burnable media is much rarer than from optical media burners from the very early 2000s. When creating backups of actual partitions they need to be intelligently split up. For a single-layer DVD, the first split needs to be 4.7GB *minus* the size of the metadata files, then subsequent splits need to be the whole 4.7GB. Rescuezilla would have to ask the user to insert the next disk (if applicable).

Thanks for the link to Parchive, there's other long-term storage archival formats like [lzip](https://en.wikipedia.org/wiki/Lzip) which may also be suitable Lzip is already fully supported by Clonezilla, and Rescuezilla v2.0+. I'll have to do more research to figure out which is the better format.

For what it's worth, I have found writeable optical media (and SD card and USB flash memory) to be a TERRIBLE for long-term cold storage of data. But I have found mechanical hard drives to be _extremely_ good. From iPods to laptops to early 1990s drives they all just keep trucking along. Maybe stay away from SMR and Helium mechanical drives as those have tight tolerances, and stick with lower RPM drives.

**Comment by shasheene** [comment]
I am closing this old GitHub issue because I don't intend to implement the ability to burn to DVD or Blu-Ray optical media from Rescuezilla itself.

Even in 2010/2011 when the feature request was made for Redo Backup and Recovery (let alone in 2019), the feature doesn't really make sense for the application.


If there is a very strong use-case, I may reconsider. But it's unlikely I'll ever implement it.
